src/mlpack/methods/ann/layer/repeat_impl.hpp

```diff
@@ -0,0 +1,204 @@
+    multiples = other.multiples;
+    interleave = other.interleave;
+    outIdxs = other.outIdxs;
+    sizeMult = other.sizeMult;
+    backIdxs = other.backIdxs;
+  }
+
+  return *this;
+}
+
+template<typename MatType>
+RepeatType<MatType>& RepeatType<MatType>::operator=(RepeatType&& other) noexcept
+{
+  if (this != &other)
+  {
+    Layer<MatType>::operator=(std::move(other));
+    multiples = std::move(other.multiples);
+    interleave = std::move(other.interleave);
+    outIdxs = std::move(other.outIdxs);
+    sizeMult = other.sizeMult;
+    backIdxs = other.backIdxs;
+  }
+
+  return *this;
+}
+
+template<typename MatType>
+void RepeatType<MatType>::ComputeOutputDimensions()
+{
+  if (multiples.size() > this->inputDimensions.size())
+  {
+    std::ostringstream oss;
+    oss << "Repeat::ComputeOutputDimensions(): multiples vector must "
+        << "have the same or fewer dimensions than InputDimensions";
+    throw std::invalid_argument(oss.str());
+  }
+
+  size_t inputSize = this->inputDimensions[0];
+  for (size_t i = 1; i < this->inputDimensions.size(); i++)
+  {
+    inputSize *= this->inputDimensions[i];
+  }
+  arma::umat idxs = arma::regspace<arma::uvec>(0, inputSize - 1);
+
+  // Here, we are going to pre-compute the source index for each output
+  // for a single tensor.  Since the tensors are flattened into 1-d
+  // vectors, we can fill the output row-wise based on these
+  // indices.
+  this->outputDimensions = this->inputDimensions;
+  sizeMult = 1;
+  size_t outSize = 1;
+
+  // iteratively reshape the index matrix such that the dimension
+  // to be repeated (and all prior) are flattened to a column, and
+  // then repelem rowwise or repmat columwise.
+  for (size_t i = 0; i < multiples.size(); i++)
+  {
+    if (multiples[i] != 1)
+    {
+      if (interleave)
+      {
+        // For the first dimension, we need to do the repelem columnwise.
+        if (i == 0)
+        {
+          idxs.reshape(outSize * this->inputDimensions[i],
+                       idxs.n_elem / (outSize * this->inputDimensions[i]));
+          idxs = repelem(idxs, multiples[i], 1);
+        }
+        else
+        {
+          idxs.reshape(outSize,
+                       idxs.n_elem / outSize);
+          idxs = repelem(idxs, 1, multiples[i]);
+        }
+      }
+      else
+      {
+        idxs.reshape(outSize * this->inputDimensions[i],
+                     idxs.n_elem / (outSize * this->inputDimensions[i]));
+        idxs = repmat(idxs, multiples[i], 1);
+      }
+      this->outputDimensions[i] *= multiples[i];
+      sizeMult *= multiples[i];
+    }
+    outSize *= this->outputDimensions[i];
+  }
+  outIdxs = idxs.as_col();
+
+  // Now, we are going to pre-compute the contribution of each output
+  // element to the input elements.  This will be used in the backward
+  // pass with a simple matrix multiplication.
+  backIdxs.set_size(inputSize, sizeMult);
+  arma::uvec counts(inputSize, arma::fill::zeros);
```

Comment:
I would make this generic as well, or at least derive the ElemeType from the MatType, no need to reserve a double col vector in the RAM if we are dealing everywhere with floats.
Check #3571 to see how these are done. There are plenty of examples there.

Great work and sorry this is taking more time to get merged than expected.